### PR TITLE
Linux porting work

### PR DESCRIPTION
**Pull request by dougvj**
Here is a first stab of implementing several backends for the linux port, most notably porting UI code to GTK+3, and abstracting the GL context creation. 

Currently the code compiles on Linux, but there are several linker errors preventing full compilation and preventing testing of the code. There is a lot of work left to be done, but it's progress.

I don't have access to a windows build environment so I need these changes tested for that platform as well.

### CODE REVIEW

File: src/xenia/base/filesystem_posix.cc [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    // TODO(benvanik): pick correct response.
+    // TODO(dougvj): in posix we would grab the errno here to generate the
+    // correct error response
+    return nullptr;
+  }
+  return std::make_unique<PosixFileHandle>(path, handle);
+}
+
+static uint64_t convertUnixtimeToWinFiletime(uint64_t unixtime) {
+    uint64_t filetime = filetime = (unixtime * 10000000) + 116444736000000000;
```
Comment (janisozaur) [inline src/xenia/base/filesystem_posix.cc]: some magic number, perhaps add a comment what it is?

**Comment by janisozaur** [comment]
https://ci.appveyor.com/project/benvanik/xenia/build/1.0.337#L432:
```
..\src\xenia\ui\gl\gl_context_win.cc(10): fatal error C1083: Cannot open include file: 'xenia/ui/gl/glx_context.h': No such file or directory [C:\dev\xenia\build\xenia-ui-gl.vcxproj]
```

**Comment by Razzile** [comment]
image


jokes aside I'm glad to see some linux work 👍

**Comment by darkf** [comment]
IMHO this should be split into two parts if possible (one for the GTK+ stuff, one for the compiling on Linux stuff), if that makes sense to do.

Secondly, your license headers have the wrong names and dates.

**Comment by darkf** [comment]
If you're going to leave some of that stuff unimplemented (like the socket stuff) you should at least make it emit an unimplemented warning.

File: src/xenia/base/filesystem_posix.cc (hunk 1 above)
Comment (dougvj) [inline src/xenia/base/filesystem_posix.cc]: done

**Comment by dougvj** [comment]
I noticed there are some build errors on wgl_context_win.cc. I will tackle that after work. I will try to setup a build environment for windows so I don't have to play the commit and pray game.

**Comment by sephiroth99** [comment]
NIce! I had started a while back, but I stopped working on it seriously. Anyway I put my stuff here https://github.com/sephiroth99/xenia/tree/linux , it's mostly dummy-ing out the stuff for it to build on Linux and with some fixes.

The linking error is an issue with premake5. For the linking to work, premake needs to add a link group for all the libraries. It added support for gcc in this commit https://github.com/premake/premake-core/pull/423 but if you use clang you need to modify premake (easy to see in that commit how to change for clang).

Anyway can't wait to try it out!

**Comment by DrChat** [comment]
For now, you can stub out most of xam_net.cc. I've got a lot of local changes there to revamp it, just never committed them yet.

File: src/xenia/ui/gl/gl_context_win.cc [hunk 2]
```diff
@@ -0,0 +1,297 @@
+/**
+ ******************************************************************************
+ * Xenia : Xbox 360 Emulator Research Project                                 *
+ ******************************************************************************
+ * Copyright 2014 Ben Vanik. All rights reserved.                             *
+ * Released under the BSD license - see LICENSE in the root for more details. *
+ ******************************************************************************
+ */
+
+#include "xenia/ui/gl/gl_context.h"
```
Comment (DrChat) [inline src/xenia/ui/gl/gl_context_win.cc]: This should be gl_context_win.h

File: src/xenia/ui/gl/gl_context_x11.cc [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+namespace xe {
+namespace ui {
+namespace gl {
+
+
+DEFINE_bool(disable_gl_context_reset, false,
+            "Do not aggressively reset the GL context (helps with capture "
+            "programs such as OBS or FRAPS).");
+DEFINE_bool(gl_debug, false, "Enable OpenGL debug validation layer.");
```
Comment (DrChat) [inline src/xenia/ui/gl/gl_context_x11.cc]: Move these to gl_context.cc and put DECLARE_bools in its header.

**Comment by Margen67** [comment]
bump

**Comment by vgturtle127** [comment]
Is GTK a good choice? I thought it was being mostly retired due to code complexity and other issues. Maybe I am wrong, but I thought that's what I heard. 😆

**Comment by John-Gee** [comment]
GTK3 is still maintained and GTK4 is getting there so no GTK is not getting retired.

**Comment by vgturtle127** [comment]
@John-Ge Alright, good to know.

**Comment by scribam** [comment]
Is anyone currently working on it ?

**Comment by dougvj** [comment]
I am still around, I have been distracted with other projects. Hopefully I will get back around to this in the next month or two.

**Comment by scribam** [comment]
Thank you for the update.
Would you mind adding the package dependencies in the .travis.yml file?

**Comment by dougvj** [comment]
xenia-app compiles and links now. There are some runtime checks that fail when creating the GTK menu bar, I will debug that sometime this week. 

I merged and rebased my changes from a year ago with the master repo, cleaning up the commit history in the process. I added some commits from @sephiroth99 who has also been working on this. I then implemented several changes including vulkan instantiation and the posix filesystem stuff. 

Right now the biggest thing that probably needs to be implemented from this point, aside from general testing and debugging, are the higher level synchronization constructs (Event, Semaphore, Timer, etc). I only stubbed those out. 

@scribam I will also update the .travis.yml stuff soon. My current build environment is not Ubuntu, but I have a VM I can use to make sure things are working properly

**Comment by tmccaff** [comment]
OpenGL is no longer supported no reason to add that.

**Comment by tmccaff** [comment]
Fact most likely OpenGL will be completely deleted soon once Vulkan issues fixed

**Pull request merged**